### ISSUE TITLE
A couple questions about the implementation.

### ISSUE THREAD

**Issue by robinydv** [liusulin/DDPD#2]
Hi, thank you for releasing the code! I have a couple of questions regarding implementation.

1) The function `corrupt_data` starting at line `279` in `train_planner.py` outputs `target_mask` which contains the indices of the tokens in the sequence which were sampled from the uniform distribution.  Some of the sampled tokens may coincide with the clean data tokens.  So `target_mask`  can contain indices of tokens that are not noisy.  

2) The loss function for training DDPD-DFM-Uni which requires only one (denoising) network is still the sum of the planner loss and denoising loss?

3) Listing 1 of Appendix D contains the sampling code for DDPD-DFM-Uni. Should the final line in the following snippet contain the constant $\frac{\alpha_t}{\alpha_t + (1 - \alpha_t)/\ S}$ in Proposition 3.1?

![image](https://github.com/user-attachments/assets/1380d865-9876-4b5e-9df1-c7fa4bd1458e)

![image](https://github.com/user-attachments/assets/5289253c-496a-433e-8e13-38a728eb0544)

**Comment by liusulin** [comment]
Thanks for the interest in our work!
1. `target_mask ` contains positions that are corrupted with a uniform random noise, which corresponds to the condition that $x_t^d$ is `Noisy`, i.e. $z_t^d=N$. Hence we use `target_mask` for training the planner. The denoiser is also trained to denoise under condition that $x_t^d$ is `Noisy`.
2. For DDPD-DFM-Uni, it is using a neural network that is trained with DFM objective to minimize cross-entropy loss on all dimensions: $\mathcal L_{\text {ce }}=\mathbb E_{p_{\text {data }}\left(x_1\right) \mathcal{U}(t ; 0,1) p_{t \mid 1}\left(x_t \mid x_1\right)}  \left[\sum_d \log p_{1 \mid t}^\theta\left(x_1^d \mid x_t\right)\right]$. Check the [DFM paper](https://arxiv.org/abs/2402.04997) for details. 
3. Listing 1 of Appendix D contains the sampling code that uses a single NN trained using DFM-Uni. There are two variants of implementing DDPD sampler for this case: 1) allowing self-loop transition (i.e. $x_t^d \to x_t^d$ same values), which corresponds to Eqn. 10; 2) no self-loop transition (i.e. only allowing $x_t^d \to j^d$ and $x_t^d \neq j^d$), then the Gillepie algorithm corresponds to the one in Listing 1. In the second case, only changing value transitions are simulated, and hence the dimension is picked accordng to `p_if_change`. Those two sampling are equivalent to each other and in experiments they have similar performance.
Let me know if you have any other questions!
